### PR TITLE
556 auto generate post descriptions

### LINKED ISSUE #556: Auto generate post descriptions
Auto generate description for posts using this [tool](https://github.com/kjaymiller/kjaymiller.com/blob/main/ollama_summarizer.py)

**Details:**
The descriptions seem like the first sentence of the content and not a dedicated line that can be used for MetaTagging as well. - I recently created a tool that would generate autodescriptions for my personal blog that could be re-used for this. Check here

### LINKED ISSUE #564: No Layout field causing issues with testing
When folks are creating blog posts, they are likely not adding all of the frontmatter content. One of the things that is commonly not added is the `layout` tag.

### THIS IS A PROBLEM

When generating the page for testing - the site applies the default tag of `article` and not `post`. This works and the site looks mostly the same but it's missing much of the components needed to make the site load properly.

with `layout` defined.

```html
<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta
      name="description"
      content="Discover the inspiring stories of Uganda&#39;s next generation African Python developers who attended PyconUG 2024 and learn about their experiences in a vibrant tech community."
    />
    <title>Black Python Devs | Black Python Devs at PyconUG 2024</title>
    <link rel="stylesheet" href="/assets/css/style.css" />
    <link rel="stylesheet" href="/assets/css/pico.min.css" />
    <link rel="stylesheet" href="/assets/css/pico.colors.min.css" />
    <link rel="stylesheet" href="/assets/css/bpd.css" />
    <link
      rel="canonical"
      href="http://localhost:56178/2024-10-25-pycon-ug-2024/"
    />
    <link
      rel="alternate"
      type="application/rss+xml"
      title="Black Python Devs"
      href="http://localhost:56178/feed.xml"
    />
    <link
      rel="icon"
      href="https://fav.farm/%E2%9C%8A%F0%9F%8F%BE"
      type="image/svg"
    />
  </head>
</html>
```

without it (it goes straight to the content)

```html
<p>Wednesday, 9th, October, 2024, was Uganda’s Independence day.</p>
```

I'm not sure why this is the case but the fix is an easy one.

\*\*ensure that `layout: post` is on all of the blog posts.

I will fix this in an upcoming PR to get tests passing but there should be a pre-commit rule to fix this.

```

```

### PR DESCRIPTION
**Pull request by kjaymiller**
- **add ollama_summarizer.py**
- **update requirements**
- **update layouts to include descriptions**
- **update posts**

## Issue Link 🔗:

**Issue**: #[Issue_Number]

## UPDATE

While trying to get tests to pass, I ran into an issue where the default post was not being applied. The most verbose fix is to auto apply it when it is missing.

I wrote about this in #564 
To fix this I made a somewhat large change.
- I added the update_layout.py cli tool that checks for a layout field and will apply it if missing (you can supply as many posts to check)
- created a custom pre-commit hook that runs update_layout on any .md files in _posts that are not updated. this works like all other pre-commit hooks in that the hook will fail, but it will fix the issue.
I updated the documentation in contributors.md

## Type of Change

- [ ] Bug fix 🐞
- [x] New feature/page
- [ ] Documentation update
- [ ] Other

## Description 📋

- **What:** Provide an overview of the issue this PR addresses. Explain the context and background information.

This PR uses the ollama_summarizer tool to check all the posts for descriptions, creating new descriptions if they currently do not exist.

It also adds the new descriptions to the meta descriptions as well as the blog lists instead of excerpts...

- **Why:** Describe why the changes are being made. Highlight key updates, new features, or bug fixes.

Excerpts were being cutoff in some posts and their HTML was showing. This tool instead can be used to quickly generate descritpions based on the contents and use AI to quickly update.

It's also easy to change the description if there is a report of a bad descriptions.

- **How:** Explain how these changes will affect the project or end-users.

This will make our social media links more dynamic and engaging and also it will give folks a summary of the content they are about to explore.

## Checklist ✅

- [x] Followed the [Code of Conduct](https://github.com/BlackPythonDevs/blackpythondevs.github.io?tab=coc-ov-file) and [Contribution Guide](https://github.com/BlackPythonDevs/blackpythondevs.github.io/blob/469afd71943738b7209bfe56d5a4e35394fbaafc/CONTRIBUTING.md)
- [x] Ran `pre-commit run --all`
- [x] All tests pass locally
- [x] Added tests (if applicable)
- [X] Documentation updated (if applicable)

## Additional Notes & Screenshots

Add any additional notes or comments that might be helpful for the reviewers.

### CODE REVIEW

**Comment by TimOsahenru** [comment]
Initially found this really insightful ... wanted to try this and get my hands messy on it 😔. Did we happen to stream this?

**Comment by kjaymiller** [comment]
> Initially found this really insightful ... wanted to try this and get my hands messy on it 😔. Did we happen to stream this?

sadly no but I did write this blog post about it late last night - https://kjaymiller.com/blog/i-made-a-precommit-hook-it-wasn-t-that-bad.html

I think the auto-generate posts thing was a temporary fix. I think long term the CLI in pre-commit is the better solution where we fail pre-commit run if there is no description.

**Comment by kjaymiller** [comment]
In the meantime. @TimOsahenru can you review the PR so we can get it pushed?

**Review by TimOsahenru — COMMENTED** [review]
Looks good to me! I noticed that some articles still don’t have an author tag. We could consider adding a default, like 'BPD,' for articles without a specified author to keep things consistent.

File: _posts/2023-10-20-djangoconus.md
Comment (TimOsahenru) [inline _posts/2023-10-20-djangoconus.md:1]: Can we have a default author for this to make things a bit consistent.

File: _posts/2024-01-25-marlene-mhangami-recognized-as-python-fellow.md
Comment (TimOsahenru) [inline _posts/2024-01-25-marlene-mhangami-recognized-as-python-fellow.md:1]: and this too

File: _posts/2024-05-22-why-we-use-colored-and-coloured.md
Comment (TimOsahenru) [inline _posts/2024-05-22-why-we-use-colored-and-coloured.md:1]: and this too... we could have "BPD" as a default author for those articles without authors

**Comment by kjaymiller** [comment]
> Looks good to me! I noticed that some articles still don’t have an author tag. We could consider adding a default, like 'BPD,' for articles without a specified author to keep things consistent.

Good news is I was the author of these posts so I can update that in a followup issue. 

I think a new issue (not to delay this PR) is to have fields that have to be filled in order for pre-commit to pass...

I think we can modify the existing `update_layout.py` (maybe rename it to `posts_validation.py`) to look for parameters and optionally set a default.

Again I think that is a different issue. In fact the only reason this was added in this issue was that I wanted to ensure that our tests were passing.

**Comment by kjaymiller** [comment]
In the meantime @TimOsahenru if it's good add your review approval to get this wrapped up.

**Review by TimOsahenru — APPROVED** [review]

**Pull request merged**